### PR TITLE
refactor(core): Deduplicate isObjectLiteral, add docs and tests

### PR DESCRIPTION
**Pull request by netroy**
## Summary
Extracted out of #12310 to address some of the comments on that PR.

## Review / Merge checklist

- [x] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md)) 
- [ ] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [x] Tests included. 
- [ ] PR Labeled with `release/backport` (if the PR is an urgent fix that needs to be backported)

### CODE REVIEW

File: packages/core/src/utils.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
+type ObjectLiteral = { [key: string]: unknown };
```
Comment (ivov) [inline packages/core/src/utils.ts:1]: 1. Does this break typings?

```suggestion
type ObjectLiteral = { [key: string | symbol]: unknown };
```

2. Might be nice to document how this differs from the `object` type?

File: packages/core/src/__tests__/utils.test.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+		['Map object', new Map(), false],
+		['Set object', new Set(), false],
+		['arrow function', () => {}, false],
+		['regular function', function () {}, false],
+		['class instance', new (class TestClass {})(), false],
+		['object with custom prototype', Object.create({ customMethod: () => {} }), false],
+		['Object.create(null)', Object.create(null), false],
+	])('should return %s for %s', (_, input, expected) => {
+		expect(isObjectLiteral(input)).toBe(expected);
+	});
```
Comment (ivov) [inline packages/core/src/__tests__/utils.test.ts:26]: Some more to consider:
- does an object with a symbol key pass?
- does a `Proxy` pass?
- does a `Promise` pass?

**Review by ivov — COMMENTED** [review]
Love it, really thorough!

File: packages/core/src/__tests__/utils.test.ts (hunk 2 above)
Comment (netroy) [reply packages/core/src/__tests__/utils.test.ts:26]: > does an object with a symbol key pass?

Yes

> does a `Proxy` pass?

We can't reliable detect proxies

> does a `Promise` pass?

Nop. I'll add a test.

**Review by ivov — APPROVED** [review]

**Comment by janober** [comment]
Got released with [`n8n@1.74.1`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.74.1)

**Comment by janober** [comment]
Got released with [`n8n@1.74.1`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.74.1)

**Comment by janober** [comment]
Got released with [`n8n@1.74.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.74.0)

**Comment by janober** [comment]
Got released with [`n8n@1.74.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.74.0)

**Pull request merged**